[margin-trim][block-layout] Content at block-start edge should have trimmed margins reflected in computed style

sammygill opened the pull request:
#### 99e30b0ca8f6729c9e28bfc1d3fa4b46cfe80a42
<pre>
[margin-trim][block-layout] Content at block-start edge should have trimmed margins reflected in computed style
<a href="https://bugs.webkit.org/show_bug.cgi?id=253606">https://bugs.webkit.org/show_bug.cgi?id=253606</a>
rdar://106452955

Reviewed by Alan Baradlay.

When we trim margins at the block-start edge, we can indicate that the
renderer has a trimmed margin by replacing the calls to setMarginBefore/After
with setTrimmedMarginForChild. This will perform the same behavior of
trimming the renderer&apos;s margins by updating its m_marginBox and also
setting the margin-trim rare data bit to indicate that the margin has
been trimmed.

In order to apply this behavior for nested block content that is at the
block-start edge of the outer block container we need to keep track of
some additional state. If a block container has block-start margin-trim
set then it will push some new state onto m_blockStartTrimming to indicate
that block-start trimming should occur and propagate this information to
its children in order to determine whether they should trim. This
new structure acts as a stack that will help nested block containers
determine if they should trim the margins at their block-start edge
based upon its containing block&apos;s trimming state and its own MarginInfo
state.

A block container will push new state onto this stack in the following
scenarios:

- If a block-container has block-start margin trim set, then it will push
some new state onto the stack (true) indicating trimming should occur
- A nested block container will check to see if its containing block has trimming
state set by checking the value at the top of the stack along with whether or
not its margins can collapse through to its containing block
    - If the containing block has trimming state and the nested child&apos;s
      margins can collapse through to the top, then the nested child will
      push its own state onto the stack to use later on as it lays out
      its children. The state will be the same as its containing block&apos;s
      so that the nested block container will only trim if it is at the
      block-start edge of the containing block that has margin-trim set.

    - If the containing block has trimming state and the nested child&apos;s
      margins *cannot* collapse through to the top, then it will push
      state onto the stack (false) to indicate that it should not
      perform any sort of trimming. This will also indicate to nested
      block containers that they should also not trim.

&lt;div id=&quot;container&quot; style=&quot;margin-trim: block&quot;&gt;
    &lt;div id=&quot;outer&quot; style=&quot;margin-block-start: 10px; border: 1px solid black; width: 50px; height: 50px; background-color: green;&quot;&gt;
        &lt;div id=&quot;inner&quot; style=&quot;margin-block-start: 10px; border: 1px solid black; width: 50px; height: 50px; background-color: blue;&quot;&gt;&lt;/div&gt;
    &lt;/div&gt;
&lt;/div&gt;

Here &quot;container,&quot; will push some state onto the margin trimming stack
to indicate that it should trim margins at the block-start edge. When
&quot;outer,&quot; goes through layout, it will see that its containing block had
set some trimming state so it will do the same. Since it has a border that
means its children&apos;s margins cannot collapse through, so it should not
trim those margins and set the state appropriately. &quot;Inner,&quot; will see
that its containing block had set some margin-trim state and will need
to do the same. Since its children&apos;s margins could collapse through, it
will use the same state as its containing block. However, since the
containing block did not perform trimming, it will also not trim when
it attempts to use this information.

In ComputedStyleExtractor, we will check to see if the renderer has its
top margin trimmed by checking to see if the rare data bit was set during
layout by using RenderBox::hasTrimmedMargin.

* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/block-container-block-start-child-with-border-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/block-container-block-start-child-with-border.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/block-container-block-start-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/block-container-block-start-self-collapsing-nested-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/block-container-block-start-self-collapsing-nested.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/block-container-block-start.html: Added.
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::rendererCanHaveTrimmedMargin):
(WebCore::isLayoutDependent):
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutBlockChildren):
(WebCore::RenderBlockFlow::layoutBlockChild):
(WebCore::RenderBlockFlow::collapseMarginsWithChildInfo):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::physicalToFlowRelativeDirectionMapping const):
(WebCore::RenderBox::hasTrimmedMargin const):
* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::isBlockLevelBox const):
* Source/WebCore/rendering/RenderLayoutState.cpp:
(WebCore::RenderLayoutState::RenderLayoutState):
* Source/WebCore/rendering/RenderLayoutState.h:
(WebCore::RenderLayoutState::RenderLayoutState):
(WebCore::RenderLayoutState::pushBlockStartTrimming):
(WebCore::RenderLayoutState::peekBlockStartTrimming):
(WebCore::RenderLayoutState::popBlockStartTrimming):

Canonical link: <a href="https://commits.webkit.org/263412@main">https://commits.webkit.org/263412@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/44919eb9143b868b5f4e3c26d3650f4a4e153381

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4519 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4638 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4783 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6004 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4682 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4508 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4769 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4594 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4928 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4584 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4689 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4050 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6010 "Built successfully") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4040 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/9027 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4038 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4112 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5644 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4495 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3668 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4029 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/4039 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1112 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8069 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4392 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.h:
```diff
@@ -181,6 +181,12 @@ class RenderBlockFlow : public RenderBlock {
+        // Used to cache the value of canTrimBlockStartMargins inside RenderLayoutState. Nested blocks
+        // could set the value in RenderLayoutState to false in order to make sure no more trimming occurs
+        // beyond that point, but we still need to make sure the margins of our children at the block-start
+        // edge have their margins trimmed.
```
Could you explain (maybe with a test case) at what point would the layout state value be different from this margin info value? It looks like you track 2 states and they may or may not have the same value and I am just curious where in block layout/margin collapsing would they diverge.

sammygill replied:
Sure, here is a small test case:

```
<div id="container" style="margin-trim: block">
    <div id="outer" style="margin-block-start: 10px; border: 1px solid black; width: 50px; height: 50px; background-color: green;">
        <div id="inner" style="margin-block-start: 10px; border: 1px solid black; width: 50px; height: 50px; background-color: blue;"></div>
    </div>
</div>
```
The outer div "container" will set the trimming layout state to true to indicate that margins at the block-start should be trimmed. Before it recurses into the layout for "outer" it will cache that state inside `MarginInfo`. "outer" will see that its block-start margins can't collapse through because of the border so it sets the value in layout state value to false. By the time we get back into container its cached value differs from the value in layout state because of this

alanbaradlay commented on Source/WebCore/rendering/RenderLayoutState.h:
```diff
@@ -111,6 +112,14 @@ class RenderLayoutState {
     void resetLeadingTrim() { m_leadingTrim = { }; }
 
+    void pushBlockStartTrimming(bool blockStartTrimming) { m_blockStartTrimming.append(blockStartTrimming); }
+    std::optional<bool> peekBlockStartTrimming() { return m_blockStartTrimming.isEmpty() ? std::nullopt : std::optional(m_blockStartTrimming.last()); }
```
blockStartTrimming() is probably closer to WebKit style and also it's a const function.

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderLayoutState.h:
```diff
@@ -111,6 +112,14 @@ class RenderLayoutState {
+    std::optional<bool> peekBlockStartTrimming() { return m_blockStartTrimming.isEmpty() ? std::nullopt : std::optional(m_blockStartTrimming.last()); }
+    void popBlockStartTrimming() 
+    {
+        ASSERT(!m_blockStartTrimming.isEmpty());
```
I bet removeLast() asserts on this already.

sammygill replied:
Yup (indirectly), changed!

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -1056,9 +1074,14 @@ void RenderBlockFlow::layoutBlockChild(RenderBox& child, MarginInfo& marginInfo,
+    if (marginInfo.atBeforeSideOfBlock() && !child.isSelfCollapsingBlock()) {
         marginInfo.setAtBeforeSideOfBlock(false);
 
+        if (auto* layoutState = frame().view()->layoutContext().layoutState(); layoutState->peekBlockStartTrimming().has_value()) {
```
null deref when no layoutState

alanbaradlay replied:
nit: operator bool calls has_value()

sammygill replied:
Changed both (along with the other calls to `has_value()` )

sammygill committed:
[margin-trim][block-layout] Content at block-start edge should have trimmed margins reflected in computed style
https://bugs.webkit.org/show_bug.cgi?id=253606
rdar://106452955

Reviewed by Alan Baradlay.

When we trim margins at the block-start edge, we can indicate that the
renderer has a trimmed margin by replacing the calls to setMarginBefore/After
with setTrimmedMarginForChild. This will perform the same behavior of
trimming the renderer's margins by updating its m_marginBox and also
setting the margin-trim rare data bit to indicate that the margin has
been trimmed.

In order to apply this behavior for nested block content that is at the
block-start edge of the outer block container we need to keep track of
some additional state. If a block container has block-start margin-trim
set then it will push some new state onto m_blockStartTrimming to indicate
that block-start trimming should occur and propagate this information to
its children in order to determine whether they should trim. This
new structure acts as a stack that will help nested block containers
determine if they should trim the margins at their block-start edge
based upon its containing block's trimming state and its own MarginInfo
state.

A block container will push new state onto this stack in the following
scenarios:

- If a block-container has block-start margin trim set, then it will push
some new state onto the stack (true) indicating trimming should occur
- A nested block container will check to see if its containing block has trimming
state set by checking the value at the top of the stack along with whether or
not its margins can collapse through to its containing block
    - If the containing block has trimming state and the nested child's
      margins can collapse through to the top, then the nested child will
      push its own state onto the stack to use later on as it lays out
      its children. The state will be the same as its containing block's
      so that the nested block container will only trim if it is at the
      block-start edge of the containing block that has margin-trim set.

    - If the containing block has trimming state and the nested child's
      margins *cannot* collapse through to the top, then it will push
      state onto the stack (false) to indicate that it should not
      perform any sort of trimming. This will also indicate to nested
      block containers that they should also not trim.

<div id="container" style="margin-trim: block">
    <div id="outer" style="margin-block-start: 10px; border: 1px solid black; width: 50px; height: 50px; background-color: green;">
        <div id="inner" style="margin-block-start: 10px; border: 1px solid black; width: 50px; height: 50px; background-color: blue;"></div>
    </div>
</div>

Here "container," will push some state onto the margin trimming stack
to indicate that it should trim margins at the block-start edge. When
"outer," goes through layout, it will see that its containing block had
set some trimming state so it will do the same. Since it has a border that
means its children's margins cannot collapse through, so it should not
trim those margins and set the state appropriately. "Inner," will see
that its containing block had set some margin-trim state and will need
to do the same. Since its children's margins could collapse through, it
will use the same state as its containing block. However, since the
containing block did not perform trimming, it will also not trim when
it attempts to use this information.

In ComputedStyleExtractor, we will check to see if the renderer has its
top margin trimmed by checking to see if the rare data bit was set during
layout by using RenderBox::hasTrimmedMargin.

* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/block-container-block-start-child-with-border-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/block-container-block-start-child-with-border.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/block-container-block-start-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/block-container-block-start-self-collapsing-nested-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/block-container-block-start-self-collapsing-nested.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/block-container-block-start.html: Added.
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::rendererCanHaveTrimmedMargin):
(WebCore::isLayoutDependent):
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutBlockChildren):
(WebCore::RenderBlockFlow::layoutBlockChild):
(WebCore::RenderBlockFlow::collapseMarginsWithChildInfo):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::physicalToFlowRelativeDirectionMapping const):
(WebCore::RenderBox::hasTrimmedMargin const):
* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::isBlockLevelBox const):
* Source/WebCore/rendering/RenderLayoutState.cpp:
(WebCore::RenderLayoutState::RenderLayoutState):
* Source/WebCore/rendering/RenderLayoutState.h:
(WebCore::RenderLayoutState::RenderLayoutState):
(WebCore::RenderLayoutState::pushBlockStartTrimming):
(WebCore::RenderLayoutState::peekBlockStartTrimming):
(WebCore::RenderLayoutState::popBlockStartTrimming):

Canonical link: https://commits.webkit.org/263412@main
